contracts/VoteRestrictionForPosting.sol

```diff
@@ -0,0 +1,196 @@
+        address indexed user,
+        uint256 indexed token,
+        bytes32 indexed postDataHash,
+        uint256 timestamp
+    );
+
+    modifier onlyKoruDao() {
+        require(
+            msg.sender == koruDao,
+            "VoteRestrictionForPosting: Only KoruDao"
+        );
+        _;
+    }
+
+    modifier onlyGelatoRelay() {
+        require(isTrustedForwarder(msg.sender), "KoruDao: Only GelatoRelay");
+        _;
+    }
+
+    constructor(
+        address _gelatoRelay,
+        uint256 _actionInterval,
+        int256 _voteThreshold,
+        address _koruDao,
+        IERC721MetaTxEnumerableUpgradeable _koruDaoNft,
+        uint256 _koruDaoProfileId
+    ) ERC2771Context(_gelatoRelay) {
+        actionInterval = _actionInterval;
+        voteThreshold = _voteThreshold;
+        koruDao = _koruDao;
+        koruDaoNft = _koruDaoNft;
+        koruDaoProfileId = _koruDaoProfileId;
+    }
+
+    function addToQueue(DataTypes.PostData calldata _postData)
+        external
+        onlyGelatoRelay
+    {
+        require(
+            _postData.profileId == koruDaoProfileId,
+            "VoteRestrictionForPosting: Only post for KoruDao"
+        );
+
+        address user = _msgSender();
+        uint256 token = getKoruDaoNftTokenId(user);
+
+        bytes32 tokenActionHash = keccak256(
+            abi.encode(token, IKoruDao.Action.POST)
+        );
+
+        // Check that NFT did not post since lastActionTime
+        checkTimeRestriction(tokenActionHash);
+
+        // Update lastActionTime
+        lastActionTime[tokenActionHash] = block.timestamp;
+
+        // Reset votes for postDataHash
+        bytes32 postDataHash = keccak256(abi.encode(_postData, user));
+        delete votes[postDataHash];
+        delete _voters[postDataHash];
+
+        // Store new post from token holder
+        postDataByTokenAction[tokenActionHash] = postDataHash;
+
+        emit LogAddToQueue(user, token, postDataHash, block.timestamp);
+    }
+
+    function checkAndUpdateRestriction(address _user, uint256 _action)
+        external
+        onlyKoruDao
+        returns (uint256 token)
+    {
+        token = getKoruDaoNftTokenId(_user);
+
+        bytes32 tokenActionHash = keccak256(abi.encode(token, _action));
+        bytes32 postDataHash = postDataByTokenAction[tokenActionHash];
+        checkVoteRestriction(postDataHash);
+
+        delete postDataByTokenAction[tokenActionHash];
+        delete votes[postDataHash];
+        delete _voters[postDataHash];
+    }
+
+    function upvote(bytes32 _postDataHash) external onlyGelatoRelay {
+        address user = _msgSender();
+        uint256 votingPower = getKoruDaoNftVotingPower(user);
+
+        if (_voters[_postDataHash].contains(user))
+            revert("VoteRestrictionForPosting: Already voted");
+
+        _voters[_postDataHash].add(user);
+        votes[_postDataHash] += int256(votingPower);
+    }
+
+    function downvoted(bytes32 _postDataHash) external onlyGelatoRelay {
```

Comment:
shouldn'T this be called `downvote`? 

Generally we could outsource most the code of up and downvote into an internal function as it is almost the same code.